<a href="https://colab.research.google.com/github/prakha10/data-visualization.github.io/blob/main/19_Missing_and_Incomplete.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Missing and Incomplete

Often datasets will be missing entries.  There are many approaches we can take to dealing with these errors and omissions.  I will examine a dataset on the characters from <ins> The Lord of The Rings</ins>

## Finding NaN's

In [ ]:
import pandas as pa

df = pa.read_csv('https://raw.githubusercontent.com/nurfnick/Data_Viz/main/lotr_characters.csv')

df.head()

birth              death  gender                             hair height  \
0      NaN                NaN  Female                              NaN    NaN   
1  TA 2978  February 26 ,3019    Male  Dark (book) Light brown (movie)    NaN   
2      NaN        March ,3019    Male                              NaN    NaN   
3   TA 280             TA 515    Male                              NaN    NaN   
4      NaN                NaN    Male                              NaN    NaN   

                     name    race  realm        spouse  
0                  Adanel     Men    NaN       Belemir  
1                 Boromir     Men    NaN           NaN  
2                  Lagduf    Orcs    NaN           NaN  
3                  Tarcil     Men  Arnor  Unnamed wife  
4  Fire-drake of Gondolin  Dragon    NaN           NaN

We see right away that there are lots of `NaN`'s.  This is an empty field in our dataset.  Some characters are mentioned but never given much more background than a name.

In [ ]:
df.isnull().sum(axis = 0)

birth     207
death     315
gender    143
hair      734
height    813
name        0
race      140
realm     714
spouse    403
dtype: int64

There are null values in every column except name.

In [ ]:
df.isnull().sum(axis = 1).value_counts().sort_index()

0     15
1     59
2    185
3    236
4    178
5     81
6     20
7      1
8    136
dtype: int64

Here we see that there are only 15 entries with all fields and 136 that are name only (since name was never blank!)  Let's look at just those characters.

In [ ]:
df[~df.isnull().any(axis = 1)]

birth  \
125                    SA 3209   
134  YT, and perhaps firstborn   
166                         YT   
186                    TA 2990   
194                     FA 532   
204                    SA 3119   
530                         YT   
551     Possibly pre First Age   
579                       3019   
620                    TA 2925   
686                    YT 1362   
692                    YT 1169   
795                  First Age   
802                    YT 1050   
873              March 1 ,2931   

                                                 death            gender  \
125                                               TA 2              Male   
134                                        Still Alive              Male   
166                                             FA 400              Male   
186                                              FO 63              Male   
194  Still alive; departed to ,Aman, on ,September ...              Male   
204                                           SA 3441               Male   
530  Still alive, departed over the sea in the earl...              Male   
551                      Unknown; possibly still alive  Most likely male   
579                                    February 293019              Male   
620                                            TA 3007              Male   
686  Still alive: Departed over the sea on ,Septemb...            Female   
692                                            YT 1497              Male   
795                       Presumably departed to ,Aman              Male   
802                                             FA 502              Male   
873                                             FO 120              Male   

                hair                           height                  name  \
125            Black             Very tall almost 7'1               Isildur   
134  Probably Golden                             Tall                 Ingwë   
166             Dark                             Tall                   Eöl   
186      Dirty blond                         Tall-6'6                  omer   
194             Dark                             Tall                Elrond   
204            Brown                           7' 10"               Elendil   
530           Silver                             Tall              Celeborn   
551             None                             Huge  Watcher in the Water   
579     Dark (movie)                    6' 6" (movie)                 Uglúk   
620     Brown (film)              1.76m / 5'9" (film)                  Bain   
686           Golden                             Tall             Galadriel   
692            Raven                             Tall                Fëanor   
795           Golden                             Tall             Thranduil   
802           Silver  Tallest of the Elven-folk, 8'2"               Thingol   
873             Dark                     198cm (6'6")    Aragorn II Elessar   

           race                              realm  \
125         Men                       Arnor,Gondor   
134       Elves                 Valinor,Taniquetil   
166       Elves                         Nan Elmoth   
186         Men                              Rohan   
194  Half-elven                          Rivendell   
204         Men                       Arnor,Gondor   
530       Elves  Eregion,Lothlórien,Caras Galadhon   
551     Urulóki                     Doors of Durin   
579    Uruk-hai                           Isengard   
620         Men                               Dale   
686       Elves  Eregion,Lothlórien,Caras Galadhon   
692       Elves                    Tirion,Formenos   
795       Elves            Woodland Realm,Mirkwood   
802       Elves                            Doriath   
873         Men      Reunited Kingdom,Arnor,Gondor   

                                  spouse  
125                         Unnamed wife  
134                         Unnamed wife  
166 

Of course we could ask for just the ones with 8 null values.

In [ ]:
df[df.isnull().sum(axis = 1) == 8].name

8                          Angrim
14                      Angelimar
17      Linda (Baggins) Proudfoot
18                 Bodo Proudfoot
40     Tanta (Hornblower) Baggins
                  ...            
886                        Andvír
891                        Amlach
904                         Aghan
905                       Agathor
907                      Aerandir
Name: name, Length: 136, dtype: object

I only included the names since the rest of the dataset was null!

Of course we can use this method to include only entries that have 4 or less null entries.

In [ ]:
df[df.isnull().sum(axis = 1) <= 4]

birth              death  gender  \
1                               TA 2978  February 26 ,3019    Male   
3                                TA 280             TA 515    Male   
5                               SA 2709            SA 2962    Male   
7                                    YT             FA 455    Male   
9                               SA 3219            SA 3440    Male   
..                                  ...                ...     ...   
903                             TA 2827            TA 2932    Male   
906                      Mid ,First Age             FA 495  Female   
908  YT during the ,Noontide of Valinor             FA 455    Male   
909                             TA 2917            TA 3010    Male   
910                     Before ,TA 1944    Late ,Third Age    Male   

                                hair height          name   race    realm  \
1    Dark (book) Light brown (movie)    NaN       Boromir    Men      NaN   
3                                NaN    NaN        Tarcil    Men    Arnor   
5                                NaN    NaN  Ar-Adûnakhôr    Men  Númenor   
7                             Golden    NaN        Angrod  Elves      NaN   
9                                NaN    NaN       Anárion    Men   Gondor   
..                               ...    ...           ...    ...      ...   
903                              NaN    NaN       Aglahad    Men      NaN   
906                              NaN    NaN         Aerin    Men      NaN   
908                           Golden    NaN        Aegnor  Elves      NaN   
909                              NaN    NaN    Adrahil II    Men      NaN   
910                              NaN    NaN     Adrahil I    Men      NaN   

                                    spouse  
1                                      NaN  
3                             Unnamed wife  
5                             Unnamed wife  
7                                 Eldalótë  
9                             Unnamed wife  
..                                     ...  
903                           Unnamed wife  
906                                 Brodda  
908  Loved ,Andreth but remained unmarried  
909                           Unnamed wife  
910                                    NaN  

[673 rows x 9 columns]

Maybe we only want the characters whose *realm* has been included.  We'll negate the `isnull()` command.

In [ ]:
df[~df.realm.isnull()]

birth        death  gender  \
3                                               TA 280       TA 515    Male   
5                                              SA 2709      SA 2962    Male   
9                                              SA 3219      SA 3440    Male   
10                                             SA 3118  Still alive    Male   
11                                             SA 2876      SA 3102    Male   
..                                                 ...          ...     ...   
890                                             TA 726       TA 946    Male   
892  Sometime during ,Years of the Trees, or the ,F...      SA 3434    Male   
898                                                NaN          NaN  Female   
900                                            TA 2544      TA 2645    Male   
901                                            TA 1330      TA 1540    Male   

    hair height          name   race      realm        spouse  
3    NaN    NaN        Tarcil    Men      Arnor  Unnamed wife  
5    NaN    NaN  Ar-Adûnakhôr    Men    Númenor  Unnamed wife  
9    NaN    NaN       Anárion    Men     Gondor  Unnamed wife  
10   NaN   Tall   Ar-Pharazôn    Men    Númenor    Tar-Míriel  
11   NaN    NaN  Ar-Sakalthôr    Men    Númenor  Unnamed wife  
..   ...    ...           ...    ...        ...           ...  
890  NaN    NaN       Amlaith    Men  Arthedain  Unnamed wife  
892  NaN    NaN         Amdír  Elves     Lórien  Unnamed wife  
898  NaN    NaN      Almarian    Men    Númenor  Tar-Meneldur  
900  NaN    NaN         Aldor    Men      Rohan  Unnamed wife  
901  NaN    NaN       Aldamir    Men     Gondor  Unnamed wife  

[197 rows x 9 columns]

## Imputing

The simplest method for filling in `NaN`s is to just place a value there.

In [ ]:
df.fillna(value = 0)

birth              death  gender  \
0                                     0                  0  Female   
1                               TA 2978  February 26 ,3019    Male   
2                                     0        March ,3019    Male   
3                                TA 280             TA 515    Male   
4                                     0                  0    Male   
..                                  ...                ...     ...   
906                      Mid ,First Age             FA 495  Female   
907                                   0                  0       0   
908  YT during the ,Noontide of Valinor             FA 455    Male   
909                             TA 2917            TA 3010    Male   
910                     Before ,TA 1944    Late ,Third Age    Male   

                                hair height                    name    race  \
0                                  0      0                  Adanel     Men   
1    Dark (book) Light brown (movie)      0                 Boromir     Men   
2                                  0      0                  Lagduf    Orcs   
3                                  0      0                  Tarcil     Men   
4                                  0      0  Fire-drake of Gondolin  Dragon   
..                               ...    ...                     ...     ...   
906                                0      0                   Aerin     Men   
907                                0      0                Aerandir       0   
908                           Golden      0                  Aegnor   Elves   
909                                0      0              Adrahil II     Men   
910                                0      0               Adrahil I     Men   

     realm                                 spouse  
0        0                                Belemir  
1        0                                      0  
2        0                                      0  
3    Arnor                           Unnamed wife  
4        0                                      0  
..     ...                                    ...  
906      0                                 Brodda  
907      0                                      0  
908      0  Loved ,Andreth but remained unmarried  
909      0                           Unnamed wife  
910      0                                      0  

[911 rows x 9 columns]

You should note right away that some of these zeros make no sense.  You might be more careful with your zeros.

In [ ]:
df.height.fillna(value = 0)

0      0
1      0
2      0
3      0
4      0
      ..
906    0
907    0
908    0
909    0
910    0
Name: height, Length: 911, dtype: object

Or you might not want to skew the average so much.  You could assign the mean if the remaining values were numerical.  Unfortuantely these are mostly strings with little hope of converting to a numerical value.

In [ ]:
df.height[~df.height.isnull()]

10                                 Tall
19                                 Tall
20     Tallest of the Elves of Gondolin
41                                 Tall
74                    Large and immense
                     ...               
831                                 8'5
850                                Tall
853                                Tall
873                        198cm (6'6")
881               As tall as a mountain
Name: height, Length: 98, dtype: object

We can also fill the empties by grabbing other values around our missing.

In [ ]:
df.height.fillna(method= 'pad')

0                        NaN
1                        NaN
2                        NaN
3                        NaN
4                        NaN
               ...          
906    As tall as a mountain
907    As tall as a mountain
908    As tall as a mountain
909    As tall as a mountain
910    As tall as a mountain
Name: height, Length: 911, dtype: object

`pad` took the last value and filled it forward.  We can also go the otherway with `bfill`

In [ ]:
df.height.fillna(method= 'bfill')

0      Tall
1      Tall
2      Tall
3      Tall
4      Tall
       ... 
906     NaN
907     NaN
908     NaN
909     NaN
910     NaN
Name: height, Length: 911, dtype: object

Filling ing my mode is a little tricky as the mode returns an array rather than a single value.  The code below changes all to *height* to the mode.

In [ ]:

df.height.transform(lambda x: x.fillna(value = x.mode()[0]))

0      Tall
1      Tall
2      Tall
3      Tall
4      Tall
       ... 
906    Tall
907    Tall
908    Tall
909    Tall
910    Tall
Name: height, Length: 911, dtype: object

## Imputing by Category

There is no quantitative data here so I actually have to work a little harder than I'd like.  If height was just a number you'd run some code like 

```
df.height.fillna(df.groupby('realm').height.transform('mean'))
```

To fill the NaNs with the mean from there group.  To deal with the categories I'll need to get the most frequent from category first.

In [ ]:
df.groupby(['race']).height.agg(pa.Series.mode)

race
Ainur                                                          Varies
Ainur,Maiar                                                        []
Balrog                                                             []
Black Uruk                                                        7'1
Dragon                                                             []
Dragons                             [As tall as a mountain, Gigantic]
Drúedain                                                        Short
Dwarf                                                              []
Dwarven                                                            []
Dwarves                           [4'5 - 5' (Estimate) , 4'5" (film)]
Eagle                                                              []
Eagles                                                             []
Elf                                                                []
Elves                                                            Tall
Elves,Maiar    

This is showing be that the most common height by each realm is mostly NaN.  We could to get rid of all that to help this imputation.

In [ ]:
dfrh = df[(~df.race.isna())&(~df.height.isna())]

dfrh.groupby(['race']).height.agg(pa.Series.mode)

race
Ainur                                                          Varies
Black Uruk                                                        7'1
Dragons                             [As tall as a mountain, Gigantic]
Drúedain                                                        Short
Dwarves                           [4'5 - 5' (Estimate) , 4'5" (film)]
Elves                                                            Tall
Ents                                                        Very tall
Ents,Onodrim                                                     15'4
Goblin,Orc                                   8,4  Body weight = 190kg
God                                                            Varies
Great Eagles                                                       30
Great Spiders                           [Enormous, Large and immense]
Half-elven                                                       Tall
Hobbit              [1.06m (3'6"), 1.17m (3'10"), 1.2m (3'11"), 1....
Hobbits        

The next line of code is not working as intended, only changing the first of each category to the mode. 

In [ ]:
df.height.fillna(df.groupby('race').height.transform(lambda s: s.mode()))

0      Tall
1       NaN
2       NaN
3       NaN
4       NaN
       ... 
906     NaN
907     NaN
908     NaN
909     NaN
910     NaN
Name: height, Length: 911, dtype: object

I believe this line of code does the same mistake but I leave it as another way to do the transformation and might be useful at some point.

In [ ]:
df.groupby('race', sort=False).height.apply(lambda x: x.fillna(value = x.mode()))



0      Tall
1       NaN
2       NaN
3       NaN
4       NaN
       ... 
903     NaN
906     NaN
908     NaN
909     NaN
910     NaN
Name: height, Length: 771, dtype: object

Below I am finally able to do the conversion.  I'll be honest in saying I don't understand why this works but the pandas methods would not allow the transformation on the entire mode.

In [ ]:
import numpy as np

df.height = df.height.fillna(df.groupby('race').height.transform(lambda x: next(iter(x.mode()), np.nan)))

df

birth              death  gender  \
0                                   NaN                NaN  Female   
1                               TA 2978  February 26 ,3019    Male   
2                                   NaN        March ,3019    Male   
3                                TA 280             TA 515    Male   
4                                   NaN                NaN    Male   
..                                  ...                ...     ...   
906                      Mid ,First Age             FA 495  Female   
907                                 NaN                NaN     NaN   
908  YT during the ,Noontide of Valinor             FA 455    Male   
909                             TA 2917            TA 3010    Male   
910                     Before ,TA 1944    Late ,Third Age    Male   

                                hair height                    name    race  \
0                                NaN   Tall                  Adanel     Men   
1    Dark (book) Light brown (movie)   Tall                 Boromir     Men   
2                                NaN    8'5                  Lagduf    Orcs   
3                                NaN   Tall                  Tarcil     Men   
4                                NaN    NaN  Fire-drake of Gondolin  Dragon   
..                               ...    ...                     ...     ...   
906                              NaN   Tall                   Aerin     Men   
907                              NaN    NaN                Aerandir     NaN   
908                           Golden   Tall                  Aegnor   Elves   
909                              NaN   Tall              Adrahil II     Men   
910                              NaN   Tall               Adrahil I     Men   

     realm                                 spouse  
0      NaN                                Belemir  
1      NaN                                    NaN  
2      NaN                                    NaN  
3    Arnor                           Unnamed wife  
4      NaN                                    NaN  
..     ...                                    ...  
906    NaN                                 Brodda  
907    NaN                                    NaN  
908    NaN  Loved ,Andreth but remained unmarried  
909    NaN                           Unnamed wife  
910    NaN                                    NaN  

[911 rows x 9 columns]

Lastly, I'll demonstrate the entire dataset transforming by the mode when grouped by race.

In [ ]:
df.fillna(df.groupby('race').transform(lambda x: next(iter(x.mode()), np.nan)))

birth              death  gender  \
0                       Late ,Third Age             FA 473  Female   
1                               TA 2978  February 26 ,3019    Male   
2                                   NaN        March ,3019    Male   
3                                TA 280             TA 515    Male   
4                                   NaN                NaN    Male   
..                                  ...                ...     ...   
906                      Mid ,First Age             FA 495  Female   
907                                 NaN                NaN     NaN   
908  YT during the ,Noontide of Valinor             FA 455    Male   
909                             TA 2917            TA 3010    Male   
910                     Before ,TA 1944    Late ,Third Age    Male   

                                  hair height                    name    race  \
0                                 Dark   Tall                  Adanel     Men   
1      Dark (book) Light brown (movie)   Tall                 Boromir     Men   
2    Grey/white strands of hair (film)    8'5                  Lagduf    Orcs   
3                                 Dark   Tall                  Tarcil     Men   
4                                  NaN    NaN  Fire-drake of Gondolin  Dragon   
..                                 ...    ...                     ...     ...   
906                               Dark   Tall                   Aerin     Men   
907                                NaN    NaN                Aerandir     NaN   
908                             Golden   Tall                  Aegnor   Elves   
909                               Dark   Tall              Adrahil II     Men   
910                               Dark   Tall               Adrahil I     Men   

                    realm                                 spouse  
0                  Gondor                                Belemir  
1                  Gondor                           Unnamed wife  
2    Moria,Mount Gundabad                                    NaN  
3                   Arnor                           Unnamed wife  
4                     NaN                                    NaN  
..                    ...                                    ...  
906                Gondor                                 Brodda  
907                   NaN                                    NaN  
908               Doriath  Loved ,Andreth but remained unmarried  
909                Gondor                           Unnamed wife  
910                Gondor                           Unnamed wife  

[911 rows x 9 columns]

This is a bit silly as the first person is dead before they are born!

## Your Turn

Check out the Air B&B dataset, https://raw.githubusercontent.com/nurfnick/Data_Viz/main/AB_NYC_2019.csv.  Examine how many entries are null.  Impute for atleast two columns that are null in an approriate fashion.

In [ ]:
df = pa.read_csv('https://raw.githubusercontent.com/nurfnick/Data_Viz/main/AB_NYC_2019.csv')

df.head()

id                                              name  host_id  \
0  2539                Clean & quiet apt home by the park     2787   
1  2595                             Skylit Midtown Castle     2845   
2  3647               THE VILLAGE OF HARLEM....NEW YORK !     4632   
3  3831                   Cozy Entire Floor of Brownstone     4869   
4  5022  Entire Apt: Spacious Studio/Loft by central park     7192   

     host_name neighbourhood_group neighbourhood  latitude  longitude  \
0         John            Brooklyn    Kensington  40.64749  -73.97237   
1     Jennifer           Manhattan       Midtown  40.75362  -73.98377   
2    Elisabeth           Manhattan        Harlem  40.80902  -73.94190   
3  LisaRoxanne            Brooklyn  Clinton Hill  40.68514  -73.95976   
4        Laura           Manhattan   East Harlem  40.79851  -73.94399   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0     Private room    149               1                  9  2018-10-19   
1  Entire home/apt    225               1                 45  2019-05-21   
2     Private room    150               3                  0         NaN   
3  Entire home/apt     89               1                270  2019-07-05   
4  Entire home/apt     80              10                  9  2018-11-19   

   reviews_per_month  calculated_host_listings_count  availability_365  
0               0.21                               6               365  
1               0.38                               2               355  
2                NaN                               1               365  
3               4.64                               1               194  
4               0.10                               1                 0